### PR TITLE
[read-fonts] Add useful helpers for reading COLR glyphs

### PR DESCRIPTION
**Pull request by dfrg**
This adds an intermediate "resolved" state for the paint graph where all values are promoted to `Fixed`, variation deltas are applied and transforms are collapsed into a single variant for each type (transform, translate, scale, rotate, skew).

The result is roughly equivalent to FreeType's [FT_COLR_Paint](https://freetype.org/freetype2/docs/reference/ft2-layer_management.html#ft_colr_paint) type. While the structure is slightly different, but the value of numeric fields _should_ be identical.

Also adds helper methods to the `Colr` table to retrieve base glyphs, layers, clip boxes, etc.

Lots of code, but the bulk of it is just mechanical type conversions.

There are no tests here yet. Opened #375 to track.

### CODE REVIEW

File: read-fonts/src/tables/colr.rs [hunk 1]
```diff
@@ -1,5 +1,696 @@
...
+/// variation deltas.
+#[derive(Clone, Debug)]
+pub struct ResolvedColorStop {
+    pub offset: Fixed,
+    pub palette_index: u16,
+    pub alpha: Fixed,
+}
+
+/// Collection of [`ColorStop`] or [`VarColorStop`].
+pub struct ColorStops<'a> {
```
Comment (anthrotype) [inline read-fonts/src/tables/colr.rs:210]: if it _either_ contains one _or_ the other, never both, should this be a enum with two variants?

File: read-fonts/src/tables/colr.rs [hunk 2]
```diff
@@ -1,5 +1,696 @@
...
+                    center_y: gradient.center_y().apply_delta(deltas[1]),
+                    start_angle: gradient.start_angle().apply_delta(deltas[2]),
+                    end_angle: gradient.end_angle().apply_delta(deltas[3]),
+                    color_stops: color_line.into(),
+                    extend,
+                }
+            }
+            Self::Glyph(glyph) => ResolvedPaint::Glyph {
+                glyph_id: glyph.glyph_id(),
+                paint: glyph.paint()?,
```
Comment (anthrotype) [inline read-fonts/src/tables/colr.rs:462]: is it intentional that this does not resolve all sub-paints recursively? probably so, but I figured I'd ask.
Is it even possible for a Rust struct (e.g. ResolvedPaint) to have a field that is of the same type as itself?

File: read-fonts/src/tables/colr.rs (hunk 1 above)
Comment (anthrotype) [reply read-fonts/src/tables/colr.rs:210]: no problems, I just asked out of curiousity (and based on my understanding that color stops can only ever be defined as either all variable or none at all). Just do whatever feels more convenient here

File: read-fonts/src/tables/colr.rs (hunk 2 above)
Comment (dfrg) [reply read-fonts/src/tables/colr.rs:462]: Yes, this is intentional. Fully resolving the graph up front would require boxing the recursive type.

The relationship between `Paint` and `ResolvedPaint` is similar to `FT_OpaquePaint` and `FT_COLR_Paint`.

File: read-fonts/src/tables/colr.rs [hunk 3]
```diff
@@ -1,5 +1,696 @@
...
+    index_map: Option<DeltaSetIndexMap<'a>>,
+    var_store: Option<ItemVariationStore<'a>>,
+    coords: &'a [F2Dot14],
+}
+
+impl<'a> ColrInstance<'a> {
+    /// Creates a new resolver for the given `COLR` table and normalized variation
+    /// coordinates.
+    pub fn new(colr: Colr<'a>, coords: &'a [F2Dot14]) -> Self {
+        let index_map = colr.var_index_map().and_then(|res| res.ok());
```
Comment (cmyr) [inline read-fonts/src/tables/colr.rs:115]: just trivia but you can also write this as `colr.var_index_map().transpose().ok().flatten()`.

**Review by cmyr — APPROVED** [review]
I trust that the actual choices in the API are directly motivated by your needs as a consumer, so I won't second guess you there (and nothing really jumped out in any case!) and I don't really have any nits either, so lgtm!

File: read-fonts/src/tables/colr.rs (hunk 3 above)
Comment (dfrg) [reply read-fonts/src/tables/colr.rs:115]: I _think_ I had the transpose version in there first and clippy yelled at me about it 🤷‍♂️

**Pull request merged**